packages/dd-trace/src/encode/agentless-ci-visibility.js

```diff
@@ -0,0 +1,245 @@
+'use strict'
+const tracerVersion = require('../../lib/version')
+const { truncateSpan, normalizeSpan } = require('./tags-processors')
+const Chunk = require('./chunk')
+const log = require('../log')
+const { AgentEncoder } = require('./0.4')
+
+const ENCODING_VERSION = 1
+
+function formatSpan (span) {
+  return {
+    type: span.type === 'test' ? 'test' : 'span',
+    version: ENCODING_VERSION,
+    content: normalizeSpan(truncateSpan(span))
+  }
+}
+
+class AgentlessCiVisibilityEncoder extends AgentEncoder {
+  constructor ({ runtimeId, service, env }) {
+    super(...arguments)
+    this._events = []
+    this.runtimeId = runtimeId
+    this.service = service
+    this.env = env
+    this._traceBytes = new Chunk()
+    this._stringBytes = new Chunk()
+    this._stringCount = 0
+    this._stringMap = {}
+
+    this.reset()
+  }
+
+  count () {
+    return this._events.length
+  }
+
+  append (trace) {
+    this._events = this._events.concat(trace)
+  }
+
+  _encodeValue (bytes, value) {
+    switch (typeof value) {
+      case 'string':
+        this._encodeString(bytes, value)
+        break
+      case 'number':
+        this._encodeNumber(bytes, value)
+        break
+      case 'object':
+        if (Array.isArray(value)) {
+          this._encodeArrayPrefix(bytes, value)
+          for (const event of value) {
+            this._encodeMap(bytes, event)
+          }
+          return
+        }
+        if (value.constructor && value.constructor.name === 'Identifier') {
+          this._encodeId(bytes, value)
+          return
+        }
+        this._encodeMap(bytes, value)
+        break
+      default:
+        // should not happen
+    }
+  }
+
+  _encodeNumber (bytes, value) {
+    if (Math.floor(value) !== value) { // float 64
+      return this._encodeFloat(bytes, value)
+    }
+    return this._encodeInteger(bytes, value)
```

Comment:
we can't reuse the 0.4 method for this. The new intake is a bit more picky about integer/floats, so for example I can't encode the `version: 1` as float